### PR TITLE
Added "choose range" command to gaming plugin

### PR DESCRIPTION
**Pull request by katycat5e**
Add a command hook for picking a random number or letter from a range, including the behavior of randint, but also including alphabetic characters and combined 0-9+A-Z. Also cleaned up choose command and allowed combining comma and 'or' delimiters when specifying options.

### CODE REVIEW

File: plugins/gaming.py [hunk 1]
```diff
@@ -116,19 +116,63 @@ def dice(text, notice):
...
+        try:
+            if choices[0].isdigit() and choices[1].isdigit():
+                # integer range
+                return str(random.randint(int(choices[0]), int(choices[1])))
+
+            if len(choices[0]) == 1 and choices[1].isalpha() and len(choices[1]) == 1:
+                # 0-9 + A-Z
+                # get the integer index of the digits/characters
+                intval_arg0 = char_val_map.get(choices[0])
+                intval_arg1 = char_val_map.get(choices[1])
```
Comment (linuxdaemon) [inline plugins/gaming.py]: This could be replaced with `int(choices[0], 36)` to parse the values as base36.

File: plugins/gaming.py [hunk 2]
```diff
@@ -116,19 +116,74 @@ def dice(text, notice):
...
+
+    if len(choices) == 2:
+        try:
+            try:
+                # attempt to parse as integer range
+                lower_bound, upper_bound = order_bounds(int(choices[0]), int(choices[1]))
+
+                return str(random.randint(lower_bound, upper_bound))
+
+            except ValueError:
```
Comment (linuxdaemon) [inline plugins/gaming.py:187]: A `try-except` should be limited to cover only the relevant code, in this case the `int()` conversions. Please split those conversions out on to their own line(s) and have the `try-except` cover those.

File: plugins/gaming.py [hunk 3]
```diff
@@ -116,19 +116,79 @@ def dice(text, notice):
...
+
+                choice = random.randint(lower_bound, upper_bound)
+
+                return str(val_char_map[choice]).upper()
+
             event.notice_doc()
             return
+        except ValueError:
+            notice("Invalid range given: {}".format(text))
+            return
```
Comment (linuxdaemon) [inline plugins/gaming.py]: What exactly is this `try-except` handling?

Comment (katycat5e) [reply plugins/gaming.py]: Thanks for pointing it out, as it seems it doesn't do anything useful anymore. I think it was an artifact from trying to catch bad int conversions.

File: plugins/gaming.py [hunk 4]
```diff
@@ -116,21 +116,90 @@ def dice(text, notice):
...
+                notice("When using letters, bounds must be within the range 0-9 + A-Z")
+                return
+
+        else:
+            # integer
+            try:
+                result = choose_integer(choices[0], choices[1])
+            except ValueError:
+                notice("For choosing a number, bounds must be valid integers")
+
```
Comment (linuxdaemon) [inline plugins/gaming.py:198]: Possible missing return? `result` won't be defined at this point, causing the fall through `return` to error

File: plugins/gaming.py [hunk 5]
```diff
@@ -116,21 +116,89 @@ def dice(text, notice):
...
+        raise ValueError
+
+    choice = random.randint(lower_bound, upper_bound)
+    return str(val_char_map[choice]).upper()
+
+
+@hook.command("chooser", "chooserange")
+def chooserange(text, event, notice):
+    """<a> to <b>|<a> <b> - picks a random integer or character from the given range"""
+    choices = re.split(r'\s*(?:to| +)\s*', text)
```
Comment (linuxdaemon) [inline plugins/gaming.py]: This should require whitespace around the `to`
```suggestion
    choices = re.split(r'\s*(?:\sto\s| +)\s*', text)
```

**Comment by linuxdaemon** [comment]
Please rebase your changes on top of the `gonzobot` branch.

`git pull --rebase https://github.com/snoonetIRC/CloudBot.git cloudbot` in to your branch

**Review by linuxdaemon — APPROVED** [review]

File: plugins/gaming.py [hunk 6]
```diff
@@ -119,21 +119,89 @@ def dice(text, notice):
     return "{} ({})".format(total, ", ".join(rolls))
 
 
+choice_split_re = re.compile(r'\s*(?:, ?or|,|or)\s*', re.I)  # regex for splitting choices for choose command
```
Comment (linuxdaemon) [inline plugins/gaming.py:122]: While testing, `choose a, bor, c d or d` returned an empty string, might want to add a basic unit test function to the plugin, like
```python
def test_regex_split():
    assert choice_split_re.split('a or b') == ['a', 'b']
    ...
```

**Review by linuxdaemon — CHANGES_REQUESTED** [review]

File: plugins/gaming.py (hunk 6 above)
Comment (linuxdaemon) [reply plugins/gaming.py:122]: If the function has the name `test_*`, the test suite will run it automatically

File: plugins/gaming.py [hunk 7]
```diff
@@ -119,21 +119,89 @@ def dice(text, notice):
...
-            event.notice_doc()
-            return
+        event.notice_doc()
+        return
 
     return random.choice([choice.strip() for choice in choices])
 
 
+# array of digits/characters indexed by int value
+val_char_map = [chr(x) for x in
```
Comment (A-UNDERSCORE-D) [inline plugins/gaming.py:140]: This is way overcomplex for what it does. try `list(string.digits + string.ascii_lowercase)` instead. You will need to import string

```
>>> [chr(x) for x in [c for r in (range(ord('0'), ord('9') + 1), range(ord('a'), ord('z') + 1)) for c in r]] == list(string.digits + string.ascii_lowercase)
True
```

File: plugins/gaming.py [hunk 8]
```diff
@@ -119,21 +119,89 @@ def dice(text, notice):
...
+def choose_alphanumeric(bound_a, bound_b):
+    """Tries to pick a value from the given 0-Z range"""
+    # get the integer index of the digits/characters
+    bound_a = int(bound_a, 36)
+    bound_b = int(bound_b, 36)
+
+    lower_bound, upper_bound = order_bounds(bound_a, bound_b)
+
+    if lower_bound < 0 or upper_bound > 36:
+        raise ValueError
```
Comment (A-UNDERSCORE-D) [inline plugins/gaming.py:171]: Add a message to your ValueError, please--Random errors with no apparent reason are impossible to debug.

File: plugins/gaming.py [hunk 9]
```diff
@@ -119,21 +119,89 @@ def dice(text, notice):
...
+
+    choice = random.randint(lower_bound, upper_bound)
+    return str(val_char_map[choice]).upper()
```
Comment (A-UNDERSCORE-D) [inline plugins/gaming.py:174]: val_char_map always contains strings. Casting is unnecessary.

File: plugins/gaming.py [hunk 10]
```diff
@@ -119,21 +119,89 @@ def dice(text, notice):
...
+
+
+@hook.command("chooser", "chooserange")
+def chooserange(text, event, notice):
+    """<a> to <b>|<a> <b> - picks a random integer or character from the given range"""
+    choices = re.split(r'(?:\s+to\s+|\s+)', text)
```
Comment (A-UNDERSCORE-D) [inline plugins/gaming.py:180]: Compile and store regexps once, don't recompile them every time.

**Review by A-UNDERSCORE-D — CHANGES_REQUESTED** [review]

**Pull request closed without merging**